src/arch/wasm/CodeGen.zig

```diff
@@ -6446,63 +6446,89 @@ fn airDivFloor(func: *CodeGen, inst: Air.Inst.Index) InnerError!void {
-        // leave on stack
-        _ = try func.binOp(lhs_res, rhs_res, ty, .div);
-        try func.addLabel(.local_tee, div_result.local.value);
-        _ = try func.cmp(lhs_res, zero, ty, .lt);
-        _ = try func.cmp(rhs_res, zero, ty, .lt);
+        // tee leaves the value on the stack and stores it in a local.
+        const quotient = try func.allocLocal(ty);
+        _ = try func.binOp(lhs_wasm, rhs_wasm, ty, .div);
+        try func.addLabel(.local_tee, quotient.local.value);
+
+        // select takes a 32 bit value as the condition, so in the 64 bit case we use eqz to narrow
+        // the 64 bit value we want to use as the condition to 32 bits.
+        // This also inverts the condition (non 0 => 0, 0 => 1), so we put the adjusted and
+        // non-adjusted quotients on the stack in the opposite order for 32 vs 64 bits.
+        if (wasm_bits == 64) {
+            try func.emitWValue(quotient);
+        }
+
+        // 0 if the signs of rhs_wasm and lhs_wasm are the same, 1 otherwise.
+        _ = try func.binOp(lhs_wasm, rhs_wasm, ty, .xor);
+        _ = try func.cmp(.stack, zero, ty, .lt);
+
         switch (wasm_bits) {
             32 => {
-                try func.addTag(.i32_xor);
                 try func.addTag(.i32_sub);
+                try func.emitWValue(quotient);
             },
             64 => {
-                try func.addTag(.i64_xor);
+                try func.addTag(.i64_extend_i32_u);
                 try func.addTag(.i64_sub);
             },
-            else => unreachable,
+            else => @panic("Unexpected wasm integer width"),
+        }
+
+        _ = try func.binOp(lhs_wasm, rhs_wasm, ty, .rem);
+
+        if (wasm_bits == 64) {
+            try func.addTag(.i64_eqz);
         }
-        try func.emitWValue(div_result);
-        // leave value on the stack
-        _ = try func.binOp(lhs_res, rhs_res, ty, .rem);
+
         try func.addTag(.select);
+
+        // We need to zero the high bits because N bit comparisons consider all 32 or 64 bits, and
+        // expect all but the lowest N bits to be 0.
+        // TODO: Should we be zeroing the high bits here or should we be ignoring the high bits
+        // when performing comparisons?
+        if (int_bits != wasm_bits) {
+            _ = try func.wrapOperand(.{ .stack = {} }, ty);
+        }
     } else {
         const float_bits = ty.floatBits(func.target);
         if (float_bits > 64) {
             return func.fail("TODO: `@divFloor` for floats with bitsize: {d}", .{float_bits});
         }
         const is_f16 = float_bits == 16;
 
-        const lhs_operand = if (is_f16) blk: {
-            break :blk try func.fpext(lhs, Type.f16, Type.f32);
-        } else lhs;
-        const rhs_operand = if (is_f16) blk: {
-            break :blk try func.fpext(rhs, Type.f16, Type.f32);
-        } else rhs;
+        const lhs_wasm = if (is_f16) try func.fpext(lhs, Type.f16, Type.f32) else lhs;
```

Comment:
Personally, I find trying to stick everything in a single line makes it more difficult to read. But I don't mind that change to ask you to change it. We can keep it and in the future see if we can make it consistent across the entire backend instead. I'd also like to suggest to keep names self-explanatory, the suffix `wasm` doesn't contribute anything to the meaning in this case over `operand` did.

Reply:
`lhs_wasm` is `lhs` widened to the wasm native type, in both the signed and float cases. Previously, the signed case used `lhs_res`, and the float case `lhs_operand`. This was inconsistent, and I didn't think the old names communicated how `lhs_res` / `lhs_operand` were different from `lhs`. I don't feel strongly about the name though, so if you find them unclear and have a better suggestion then I will change them.

Reply:
I find using a labeled block and break with value when the contents of the block is a single expression unnecessarily noisy given that `if .. else` has an expression form, hence the change. Whether the `if .. else` is formatted over multiple lines is a separate issue to whether the block or expression form is used (see line 6460 of the same file). Also note that my hand was forced a bit with the formatting, as `zigfmt` is fussy and will not, for example, let you split the `if .. else` expression over two lines.

This is all minor style stuff though - I don't mind changing it.

Reply:
Thank you for the explanation, that makes sense with regard to the naming of the local. As for the block, I'm heavily biased as I wrote that piece of code, but if this means a contributor finds it more difficult to read, it's a worthwhile change to me. Just for your information, `zig fmt` will accept the following:
```zig
const lhs_wasm = if (is_f16) 
    try func.fpext(lhs, Type.f16, Type.f32) 
else 
    lhs;
```
Though I'm not sure if this makes it any more readable.
Let's keep both changes you've done.